blake2b/blake2b.go

```diff
@@ -333,15 +333,14 @@ func (d *Digest) finalize(out []byte) error {
 
 	dCopy.compress()
 
-	// extract output
-	putU64LE(out[0*8:], dCopy.h[0])
-	putU64LE(out[1*8:], dCopy.h[1])
-	putU64LE(out[2*8:], dCopy.h[2])
-	putU64LE(out[3*8:], dCopy.h[3])
-	putU64LE(out[4*8:], dCopy.h[4])
-	putU64LE(out[5*8:], dCopy.h[5])
-	putU64LE(out[6*8:], dCopy.h[6])
-	putU64LE(out[7*8:], dCopy.h[7])
+	var shift uint
+	var mask uint64
+
+	for offset := 0; offset < len(out); offset++ {
+		shift = 8 * (uint(offset) % 8)
+		mask = uint64(0xFF << shift)
+		out[offset] = byte((dCopy.h[offset/8] & mask) >> shift)
+	}
```

Comment:
Sadly this is going to affect performance :(
https://github.com/golang/go/issues/40704

Reply:
Ugh, yes it does, and it's fairly harsh. But since `finalize` is only called once per `Sum` I think correctness wins over performance here. It won't affect any longer series of writes and compressions, and rough benchmarking confirmed the expected result that this mostly hurts very small write tasks where the finalization dominates:

```
$ benchstat not-fixed fixed 
name          old time/op    new time/op    delta
Hash8Bytes-4     361ns ± 0%     459ns ± 1%  +27.33%  (p=0.000 n=9+10)
Hash1K-4        1.77µs ± 0%    1.87µs ± 0%   +5.62%  (p=0.000 n=10+10)
Hash8K-4        12.9µs ± 1%    13.0µs ± 1%   +0.74%  (p=0.001 n=9+10)

name          old speed      new speed      delta
Hash8Bytes-4  22.2MB/s ± 0%  17.4MB/s ± 1%  -21.46%  (p=0.000 n=9+10)
Hash1K-4       578MB/s ± 0%   547MB/s ± 0%   -5.32%  (p=0.000 n=10+10)
Hash8K-4       633MB/s ± 1%   629MB/s ± 1%   -0.73%  (p=0.001 n=9+10)
```

Unfortunately hashing small amounts of data (say, 32 bytes of key) is my main use case. So I'd happily take a PR with a better solution, but I figured "panics on totally reasonable inputs to a top level API" needed to be fixed :)

Reply:
Overall, I think if anyone needs a **lot** more throughput from this, the most expedient solution would be the assembly from `x/crypto`. I dislike that, but it will likely double the performance.